### PR TITLE
Add batch submission function

### PR DESCRIPTION
**Pull request by dnewh**
This adds a batch submission function to saplingjs. This function takes
a url to the endpoint and the batch bytes as a payload. It also adds a
http request convenience function for making other requests.

Signed-off-by: Davey Newhall <newhall@bitwise.io>

### CODE REVIEW

File: canopy/saplingjs/src/submitter.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+export async function http(
+  method: string,
+  url: string,
+  data: Uint8Array,
+  headerFn: (request: XMLHttpRequest) => void
+): Promise<string> {
+  return new Promise((resolve, reject) => {
+    const request = new XMLHttpRequest();
+    request.open(method, url);
```
Comment (dplumb94) [inline canopy/saplingjs/src/submitter.ts]: check if this is a valid method or restrict with a string literal

File: canopy/saplingjs/src/submitter.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+      }
+    };
+    request.onerror = (): void => {
+      reject(
+        Error(
+          'The server has encountered an error. Please contact the administrator.'
+        )
+      );
+    };
+    request.send(data);
```
Comment (dplumb94) [inline canopy/saplingjs/src/submitter.ts]: What happens if the method doesn't take a payload? For instance if you are using this for a GET

File: canopy/saplingjs/src/index.ts [hunk 3]
```diff
@@ -14,6 +14,8 @@
  * limitations under the License.
  */
 
+export { http, submitBatch } from './submitter';
```
Comment (dplumb94) [inline canopy/saplingjs/src/index.ts]: I don't think we should expose `http`.  This should just be a helper function for other things in this library.

File: canopy/saplingjs/src/submitter.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  errorMessage: string;
+  errorData: number[];
+}
+
+interface BatchInfo {
+  id: string;
+  status: BatchStatus;
+}
+
+export async function http(
```
Comment (peterschwarz) [inline canopy/saplingjs/src/submitter.ts]: Could you write some documentation for these functions?

File: canopy/saplingjs/src/submitter.ts (hunk 2 above)
Comment (dnewh) [reply canopy/saplingjs/src/submitter.ts]: For `GET` and `HEAD` requests, the body/data argument is just ignored: https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/send

File: canopy/saplingjs/src/submitter.ts (hunk 1 above)
Comment (dnewh) [reply canopy/saplingjs/src/submitter.ts]: Added check. Now rejects with error message if not valid.

File: canopy/saplingjs/src/submitter.ts (hunk 4 above)
Comment (dnewh) [reply canopy/saplingjs/src/submitter.ts]: added docstrings

File: canopy/saplingjs/src/index.ts (hunk 3 above)
Comment (dnewh) [reply canopy/saplingjs/src/index.ts]: no longer exported

File: canopy/saplingjs/src/submitter.ts (hunk 2 above)
Comment (dplumb94) [reply canopy/saplingjs/src/submitter.ts]: But it is required in the function signature

Comment (dnewh) [reply canopy/saplingjs/src/submitter.ts]: added none option

File: canopy/saplingjs/src/submitter.ts [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+/**
+ * Submits a single batch of transactions
+ * @param {string}      url   The endpoint to submit the batch to
+ * @param {Uint8Array}  batch The serialized batch of transactions
+ */
+export async function submitBatch(
+  url: string,
+  batch: Uint8Array
```
Comment (peterschwarz) [inline canopy/saplingjs/src/submitter.ts]: Late to the realization that this needs to be a batch list.  That doesn't change the type but the name and documentation.

Or, the content needs to be added to a BatchList protobuf instance before it is submitted

Comment (dnewh) [reply canopy/saplingjs/src/submitter.ts]: name/docs updated

**Pull request merged**